### ISSUE TITLE
CsvWriter.WriteHeader<dynamic> buggy

### ISSUE THREAD

**Issue by jzabroski** [JoshClose/CsvHelper#1417]
**Describe the bug**
If each instance of IExpandoObject contains a different set of properties, WriteHeader uses the first object in the collection to write the header.  There also doesn't seem to be a way to say, "here is a string array of column headers".

**To Reproduce**
```c#
dynamic expando1 = new ExpandoObject();
var p = expando1 as IDictionary<String, object>;

p["A"] = "New val 1";
p["B"] = "New val 2";

dynamic expando2 = new ExpandoObject();
var q = expando2 = new ExpandoObject();

q["A"] = "New val 1";
q["B"] = "New val 2";
q["C"] = "Independent second column";

var listOfObjects = new List<dynamic>();
listOfObjects.Add(p);
listOfObjects.Add(q);

using (var memoryStream = new MemoryStream())
{
	using (var csv = new CsvHelper.CsvWriter(new StreamWriter(memoryStream)))
	{
		csv.WriteRecords(listOfTags);
	}
}
```

**Expected behavior**
I need to write an assert statement to make this clearer, but I hope my description is sufficient.

**Screenshots**

**Additional context**
Was hoping to use CsvHelper to write a FIX log parser that dumps Financial Interchange messages to Excel.

**Comment by jzabroski** [comment]
Note to self: improve test using examples from Josh given here: https://github.com/JoshClose/CsvHelper/blob/b74a2f95a101158f4cdedd25fae6e8392b58855b/src/CsvHelper.Tests/Writing/DynamicTests.cs#L50

**Comment by jzabroski** [comment]
The other problem seems to be that, even if you use WriteDynamicHeader() to write an expandoObject that contains all the headers, if some of the expandos in WriteRecords() don't contain all the elements in the header, you end up messing up the line endings.

**Comment by jzabroski** [comment]
I may also be hitting CsvWriter.DynamicPropertySort behavior.

**Comment by jzabroski** [comment]
Yeah, it seems if DynamicPropertySort behavior is not over-written, the behavior is non-deterministic.

**Comment by APIWT** [comment]
> The other problem seems to be that, even if you use WriteDynamicHeader() to write an expandoObject that contains all the headers, if some of the expandos in WriteRecords() don't contain all the elements in the header, you end up messing up the line endings.

Hmm. Do you mean that you get something like this?
```
col1,col2,col3
val1,val2
val1,val2,val3
```

Or something different?

**Comment by jzabroski** [comment]
I was trying to dump Financianl Information eXchange protocol messages to a CSV as it is easier to read many of these messages if they're in table/blotter format instead of raw format.  Inbound messages have a different set of headers than outbound messages but there can be some basic overlap like Tag150.

Outbound, you might specify custom tags between 5000-9000 for specific order instructions like algorithmic trading parameters.

Inbound, you might get tags like AvgPx (Average Price) which are execution-level information as opposed to order-level information.

**Comment by jzabroski** [comment]
I was thinking this problem can also happen if you have an object model that has polymorphic records.

e.g. suppose you want to serialize (this is not a real world model, I'm just making something up off the top of my head):

```c#
class Invoice { long Id {get;set;} ICollection<ILineItem> {get;set;} }
interface ILineItem { }
class PassthroughExpenseLineItem : ILineItem { decimal amount {get;set; } }
class MarkupTermsLineItem : ILineItem { float markupPercentage {get; set;} }
```
 I suspect if you try to serialize this using dynamic, it will fail.
